dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryStore.cs

```diff
@@ -0,0 +1,320 @@
+        {
+            if (!responseMessage.IsSuccessStatusCode)
+            {
+                this._log?.LogWarning("Failed to get item {0} from collection {1} with status code {2}", key, collectionName, responseMessage.StatusCode);
+                return null;
+            }
+
+            using (responseMessage.Content)
+            {
+                CosmosDBMemoryRecord record;
+
+                if (typeof(Stream).IsAssignableFrom(typeof(CosmosDBMemoryRecord)))
+                {
+                    record = ((CosmosDBMemoryRecord)(object)responseMessage.Content);
+                }
+                else
+                {
+                    record = await System.Text.Json.JsonSerializer.DeserializeAsync<CosmosDBMemoryRecord>(responseMessage.Content!, cancellationToken: cancel);
+                }
+
+                var embeddingHost = JsonConvert.DeserializeAnonymousType(
+                    record!.EmbeddingString,
+                    new { Embedding = new { vector = new List<float>() } });
+
+                return MemoryRecord.FromJson(
+                    record.MetadataString,
+                    new Embedding<float>(embeddingHost.Embedding.vector),
+                    record.Id,
+                    record.Timestamp);
+            }
+        }
+    }
+
+    /// <inheritdoc/>
+    public async IAsyncEnumerable<MemoryRecord> GetBatchAsync(string collectionName, IEnumerable<string> keys,
+        [EnumeratorCancellation] CancellationToken cancel = default)
+    {
+        foreach (var key in keys)
+        {
+            var record = await this.GetAsync(collectionName, key, cancel);
+
+            if (record != null)
+            {
+                yield return record;
+            }
+        }
+    }
+
+    /// <inheritdoc />
+    public async Task<string> UpsertAsync(string collectionName, MemoryRecord record, CancellationToken cancel = default)
+    {
+        record.Key = this.ToCosmosFriendlyId(record.Metadata.Id);
+
+        var entity = new CosmosDBMemoryRecord
+        {
+            CollectionId = collectionName,
+            Id = record.Key,
+            Timestamp = record.Timestamp,
+            EmbeddingString = string.Join(", ", record.Embedding.AsReadOnlySpan().ToArray()),
+            MetadataString = record.GetSerializedMetadata()
+        };
+
+        var container = this._database.Client.GetContainer(this._databaseName, collectionName);
+
+        await container.UpsertItemAsync(entity, cancellationToken: cancel, requestOptions: new ItemRequestOptions()
+        {
+            EnableContentResponseOnWrite = false,
+        });
+
+        return record.Key;
+    }
+
+    /// <inheritdoc/>
+    public async IAsyncEnumerable<string> UpsertBatchAsync(string collectionName, IEnumerable<MemoryRecord> records, CancellationToken cancel = default)
```

Comment:
Async-iterator 'CosmosDBMemoryStore.UpsertBatchAsync(string, IEnumerable<MemoryRecord>, CancellationToken)' has one or more parameters of type 'CancellationToken' but none of them is decorated with the 'EnumeratorCancellation' attribute, so the cancellation token parameter from the generated 'IAsyncEnumerable<>.GetAsyncEnumerator' will be unconsumed